### PR TITLE
[SPARK-5578][SQL][DataFrame] Provide a convenient way for Scala users to use UDFs

### PR DESCRIPTION
**Pull request by rxin**
A more convenient way to define user-defined functions.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 1]
```diff
@@ -132,14 +132,14 @@ class LogisticRegressionModel private[ml] (
   override def transform(dataset: DataFrame, paramMap: ParamMap): DataFrame = {
     transformSchema(dataset.schema, paramMap, logging = true)
     val map = this.paramMap ++ paramMap
-    val scoreFunction: Vector => Double = (v) => {
+    val scoreFunction = defineUDF((v: Vector) => {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: API-wise, is it better to just call it `UDF` instead of `defineUDF`?

Comment (rxin) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: i'm worried about name collision here ...

File: sql/core/src/main/scala/org/apache/spark/sql/Dsl.scala [hunk 2]
```diff
@@ -213,188 +211,189 @@ object Dsl {
   }
   }
   */
+
   /**
    * Call a Scala function of 0 arguments as user-defined function (UDF), and automatically
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/Dsl.scala]: Update doc strings.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 3]
```diff
@@ -132,14 +132,14 @@ class LogisticRegressionModel private[ml] (
   override def transform(dataset: DataFrame, paramMap: ParamMap): DataFrame = {
     transformSchema(dataset.schema, paramMap, logging = true)
     val map = this.paramMap ++ paramMap
-    val scoreFunction: Vector => Double = (v) => {
+    val scoreFunction = udf((v: Vector) => {
       val margin = BLAS.dot(v, weights)
       1.0 / (1.0 + math.exp(-margin))
-    }
+    } : Double)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: About the syntax, I like the following better

~~~
val margin = udf { v: Vector => 
  val margin = BLAS.dot(v, weights)
  1.0 / (1.0 + math.exp(-margin))
}
~~~

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -26,9 +27,8 @@ import org.apache.spark.sql.catalyst.errors.TreeNodeException
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.types._
 
-/* Implicits */
 import org.apache.spark.sql.TestData._
-import org.apache.spark.sql.test.TestSQLContext._
+import org.apache.spark.sql.test.TestSQLContext.{udf => _, _}
```
Comment (mengxr) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Maybe it is useful to leave a comment here about what it does.

**Comment by rxin** [comment]
Jenkins, retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 4 above)
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: We use this quite a bit in SQL tests though. Really, we should avoid wildcard important. That's a future TODO.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 3 above)
Comment (rxin) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: will do in a separate pr
